File: 4_Dicas/4.1_OpenCV/Ex11/Faces.py
import cv2, numpy, sys

def detectAndDisplay(image, faceCascade, eyesCascade):
	gray  = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
	gray = cv2.equalizeHist(gray)
	faces = faceCascade.detectMultiScale(gray)
	for (x, y, w, h) in faces:
		cv2.ellipse(image, (int(x+w/2), int(y+h/2)), (int(w/2),int(h/2)), 0, 0, 360, ( 255, 0, 255 ), 4 )
		faceROI = gray[y:y+h,x:x+h]
		eyes = eyesCascade.detectMultiScale(faceROI)
		for (xe, ye, we, he) in eyes:
			cv2.circle(image, (int(x+xe+we/2), int(y+ye+he/2)), round((we + he)*0.25), ( 255, 0, 0 ), 4 )
	cv2.imshow("Faces found", image)
	cv2.waitKey(0)

if len(sys.argv)<4:
	print("Modo de uso: " + sys.argv[0] + " Nome_Imagem Cascade_Face Cascade_Olhos")
	sys.exit(-1)

face_cascade = cv2.CascadeClassifier()
if not face_cascade.load(sys.argv[2]):
	print("--(!)Error loading face cascade")
	sys.exit(-1)

eyes_cascade = cv2.CascadeClassifier()
if not eyes_cascade.load(sys.argv[3]):
	print("--(!)Error loading eyes cascade")
	sys.exit(-1)

detectAndDisplay(cv2.imread(sys.argv[1]), face_cascade, eyes_cascade)

File: 4_Dicas/4.1_OpenCV/Ex11/Faces.cpp
#include "opencv2/objdetect.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/videoio.hpp"
#include <iostream>

using namespace std;
using namespace cv;

/** Function Headers */
void detectAndDisplay( Mat frame );

/** Global variables */
CascadeClassifier face_cascade;
CascadeClassifier eyes_cascade;

/** @function main */
int main( int argc, const char** argv )
{
	if(argc != 4)
	{
		printf("Modo de uso: %s Nome_Imagem Cascade_Face Cascade_Olhos\n",
			argv[0]);
		return -1;
	}
	//-- 1. Load the cascades
	if(!face_cascade.load(argv[2]))
	{
		cout << "--(!)Error loading face cascade\n";
		return -1;
	}
	if(!eyes_cascade.load(argv[3]))
	{
		cout << "--(!)Error loading eyes cascade\n";
		return -1;
	}
	detectAndDisplay(imread(argv[1]));
	return 0;
}

/** @function detectAndDisplay */
void detectAndDisplay( Mat frame )
{
	Mat frame_gray;
	cvtColor( frame, frame_gray, COLOR_BGR2GRAY );
	equalizeHist( frame_gray, frame_gray );

	//-- Detect faces
	std::vector<Rect> faces;
	face_cascade.detectMultiScale( frame_gray, faces );

	for ( size_t i = 0; i < faces.size(); i++ )
	{
		Point center( faces[i].x + faces[i].width/2, faces[i].y + faces[i].height/2 );
		ellipse( frame, center, Size( faces[i].width/2, faces[i].height/2 ), 0, 0, 360, Scalar( 255, 0, 255 ), 4 );

		Mat faceROI = frame_gray( faces[i] );

		//-- In each face, detect eyes
		std::vector<Rect> eyes;
		eyes_cascade.detectMultiScale( faceROI, eyes );

		for ( size_t j = 0; j < eyes.size(); j++ )
		{
			Point eye_center( faces[i].x + eyes[j].x + eyes[j].width/2, faces[i].y + eyes[j].y + eyes[j].height/2 );
			int radius = cvRound( (eyes[j].width + eyes[j].height)*0.25 );
			circle( frame, eye_center, radius, Scalar( 255, 0, 0 ), 4 );
		}
	}

	//-- Show what you got
	imshow( "Capture - Face detection", frame );
	waitKey(0);
}